WIP: Feature/ast detection transformation

marionebl opened the pull request:
- [x] Inject Globals
- [x] Semi-bundle mode for markup generation
- [x] Double-check implicit imports
- [x] Double-check plain detection
- [x]  Teach react-to-markup to resolve its dependencies. See https://github.com/sinnerschrader/patternplate/issues/53

marionebl committed:
feat(transformation): bootstrap ast-based react transformation

marionebl committed:
feat(detection): resolve transitive demo dependencies properly

ZauberNerd committed:
fix(system): add / update missing dependencies

ZauberNerd committed:
fix(detection): check for existence of default export

marionebl committed:
chore(system): update jsonlint usage

marionebl committed:
chore(system): update commit linter usage

marionebl committed:
fix(generation): push npm resolvable names to file dependencies

marionebl committed:
fix(detection): avoid babylon bug by removing jsx at the end

marionebl committed:
fix(generation): require react on top-level

marionebl committed:
chore(system): tend devDependencies

ZauberNerd committed:
fix(generation): only include react import if it is missing

Separate the AST paths for import / require calls from their identifiers
for easier processing and filtering

marionebl committed:
Merge pull request #3 from nerdlabs/fix-import-detection

fix(generation): only include react import if it is missing

marionebl committed:
chore(system): streamline watch task

marionebl committed:
fix(generation): fix mtime lookup

marionebl committed:
test(system): add basic test setup

marionebl committed:
fix(generation): rewrite props deref only for stateless components

marionebl committed:
style(transformation): describe steps for react components creation

marionebl commented:
So apparently the `transforms.react.global.<foo>` injections are pretty hard to implement properly for all cases which are

* user-written stateless components
* user-written React classes
* inferred stateless components
* inferred React classes
* other blocks/scopes in the jsx files

marionebl committed:
feat(generation): enable partial global support

marionebl committed:
feat(generation): implement legacy global injection support

marionebl committed:
feat(detection): throw better parser errors

marionebl committed:
feat(detection): allow trailing function commas

marionebl commented:
I'd consider this feature complete but holding it back to investigate:

* [ ] An issue with stanzas in the form of `this.props.someArray = this.props.someArray || ['foo', 'bar', 'baz']` in plain jsx files
* [ ] An issue with plain jsx files without semicolon on the last js line before jsx failing to parse 

Furthermore I'd like to discuss if we should deprecate `global` altogether or rather the convoluted access via `this`. The deprecation warnings are currently commented, deprecation warnings for access via `this.<contextKey>` are in place, instructing users to use `global.<contextKey>` instead.

cc @ZauberNerd @KnisterPeter @Dangoo

KnisterPeter commented on source/create-react-component.js:
```diff
@@ -0,0 +1,123 @@
+	// Inject globals into ast
+	injectGlobals(ast, globals);
+
+	// If a default export OR plain jsx was found, assume we deal with
```
I think you mean 'no plain jsx' was found

KnisterPeter commented on source/supports-stateless-components.js:
```diff
@@ -0,0 +1,7 @@
+import compareSemver from 'compare-semver';
+
+export default function supportsStatelessComponents(React) {
+	return compareSemver.gt(React.version, ['0.13.3']);
```
If possible it would be better to use '>= 0.14'. It would be more readable.

KnisterPeter commented on test/_mocks.js:
```diff
@@ -0,0 +1,87 @@
+	dependencies: {}
+};
+
+export const statelessFile = {
```
Name this statelessComponent or statelessFunction

KnisterPeter commented on test/_mocks.js:
```diff
@@ -0,0 +1,87 @@
+	dependencies: {}
+};
+
+export const fullFile = {
```
Name this reactComponent or event better reactClass

KnisterPeter commented on test/_mocks.js:
```diff
@@ -0,0 +1,87 @@
+	`)),
+	path: 'full/index.jsx',
+	dependencies: {}
+};
```
Add a test case for React.createClass...

KnisterPeter commented:
I'm not sure what you mean with 'this.<context>' vs 'global.<context>'.
I would not even instruct people to use context or create something using it. It should be considered a last resort for some problems, but in most cases it is just implicit API.

Please elaborate on this a but more and give an example.

marionebl commented:
Expanding on the context access discussion; I'll assume the users have decided to use global variables for some reason.

As currently implemented `patternplate-transform-react` makes the values found in 
```js
// environment config or sth.
{
   "global": {
      "foo": "bar"
   }
}
```

available via 

```js
// some jsx file
console.log(this.foo); // => "bar"
```

This has some obvious hazards, such as overriding of / collisions with React builtins, as seen in some of our projects:

```js
// config
{
   "global": {
      "context": {
         "foo": "bar"
      }
   }
}
// usage; blows in React 0.14 (this.context is non-configurable!), works fine in React 0.13
console.log(this.context); // => {"foo": "bar"}
```
I propose to support the access to `this.<globalKey>` (https://github.com/sinnerschrader/patternplate-transform-react/pull/1/files#diff-79d7d5bfd5e055c7cacde4fbc12f8f6aR28) until `0.4.0` of `patternplate-transform-react`,  and show pretty harsh deprecations (https://github.com/sinnerschrader/patternplate-transform-react/pull/1/files#diff-c60c67bd505d9618ebf77a7f130903d0R66).

KnisterPeter commented:
Okay, got it. Sounds reasonable. But I think there don't need to be any reason to use globals.
It would be more wise to redirect people to use react context (I know what I said above) as last resort before using globals.

KnisterPeter commented:
And even worse: If someone whats to use pattern plate on node side overwriting 'gobal' is at least as bad as overwriting some of react internals (e.g. this.context...)

marionebl commented:
A major issue with existing code bases using the current transform:

babel won't parse 

```js
var foo = bar

<div />
```

as opposed to (note the semicolon)

```js
var foo = bar;

<div />
```

I am not quite sure on how to tackle this in a backwards-compatible fashion. Any suggestions @ZauberNerd @KnisterPeter. 

Additionally this error is really hard to miss for the users – how should we handle this in the future? 

Do we help with this error, leave it to user land, change our semantics to make it less likely for users to run into?

KnisterPeter commented:
This is not an easy one for sure.
I think it would be best to report the error to the user, but it need to be understandable.
It introduces a breaking change with this transform, but I would guess it is an easy one.

marionebl commented:
Simplified possible solution paths for this

1. Detect the error
    - `babylon` throws upon parsing
    - search for first line matching `\s<`
    - get the last line with content before this, check if it ends with an `;`
    - if it ends with `;`, throw the original error - else continue

2. Handle the error, we should decide if we want to do (a), (b) or (c)
2a. Throw an enhanced error message
2b. Log a deprecation and fix the syntax error, continue with the transform
2c. Fix the syntax error without logging a deprecation

My thoughts on this are:
(a) is the safest route, there already is an exception thrown by `babylon` - we just add more information to it. In the worst case the detection of this specific error is faulty and we add a tiny bit of overhead to the debugging process. This could be alleviated by retaining the original error message and phrasing our bit defensively. **Users that omitted the semicolon will have to take immediate action after updating to the new transform.**

(b) is risky. If we mess up the detection we defer the exception to a later step and possibly hide the original babylon exception from the user, making debugging pretty hard. In the best case users get time to handle to adapt according to our deprecation warning and we can remove the extra code path in the future. **No immediate action by users required**

(c) is very risky. We establish extra syntax in the long run. **No immediate action by users required**


I'm leaning towards (b), possibly retaining (a) long-term to help with debugging this particular programming mistake.

KnisterPeter commented:
I would take (a).

marionebl commented:
This one is important. Any thoughts @Dangoo @martinemmert @ZauberNerd ?

marionebl commented:
cc @LKuechler @vegiepatch

LKuechler commented:
(a) sounds to be the safest way to go.
That the user has to add a omitted semicolon after an update should not be a problem when there is a good/understandable error message.

ZauberNerd commented:
Yes, as we already talked about: Go with (a) and explain why ASI shit is ;)

marionebl committed:
chore(system): test flowtype

marionebl commented:
Code in lieu of 

```js
// patterns/foo/index.jsx
console.log()
// foo bar
<div/>
```

will give you the following error:
```
Detected possible missing semi-colon before plain jsx block:

Unexpected token (4:5)

  1 | 
> 2 | console.log()
    |              ^
  3 | // foo bar
  4 | <div/>
  5 | 

You can fix this by adding a semi-colon at line 2:
console.log();

or wrapping jsx in parenthesis, e.g.:
(<div />)
```

Clear enough @KnisterPeter @Dangoo @ZauberNerd @LKuechler?

marionebl committed:
chore(system): update babel-code-frame

marionebl committed:
chore(system): update ava

marionebl committed:
feat(detection): enhance error messages for asi errors in plain jsx

marionebl commented:
Remaining todos:

- [ ] Retain references to static properties on react classes
- [ ] Avoid naming collisions with parameters when creating stateless components

marionebl committed:
fix(generation): avoid naming collisions during transformation

marionebl committed:
test(system): beef up test suite

marionebl committed:
chore(system): add dependency checker utils

marionebl committed:
style(detection): remove stray debug comments

marionebl committed:
style(detection): remove stray console.log

marionebl committed:
fix(detection): avoid shadowing parser errors

marionebl committed:
chore(release): 0.3.0
<a name="0.3.0"></a>
# [0.3.0](https://github.com/sinnerschrader/patternplate-transform-react/compare/v0.2.3...v0.3.0) (2016-04-16)

### Bug Fixes

* **detection:** avoid babylon bug by removing jsx at the end ([74a8006](https://github.com/sinnerschrader/patternplate-transform-react/commit/74a8006))
* **detection:** avoid shadowing parser errors ([795ca69](https://github.com/sinnerschrader/patternplate-transform-react/commit/795ca69))
* **detection:** check for existence of default export ([c093fcb](https://github.com/sinnerschrader/patternplate-transform-react/commit/c093fcb))
* **generation:** avoid naming collisions during transformation ([40e29b0](https://github.com/sinnerschrader/patternplate-transform-react/commit/40e29b0))
* **generation:** fix mtime lookup ([df4c8ce](https://github.com/sinnerschrader/patternplate-transform-react/commit/df4c8ce))
* **generation:** only include react import if it is missing ([244fc6c](https://github.com/sinnerschrader/patternplate-transform-react/commit/244fc6c))
* **generation:** push npm resolvable names to file dependencies ([39c138f](https://github.com/sinnerschrader/patternplate-transform-react/commit/39c138f))
* **generation:** require react on top-level ([ad35e45](https://github.com/sinnerschrader/patternplate-transform-react/commit/ad35e45))
* **generation:** rewrite props deref only for stateless components ([2e5bf3a](https://github.com/sinnerschrader/patternplate-transform-react/commit/2e5bf3a))
* **system:** add / update missing dependencies ([bd5fc13](https://github.com/sinnerschrader/patternplate-transform-react/commit/bd5fc13))

### Features

* **detection:** allow trailing function commas ([9d1804c](https://github.com/sinnerschrader/patternplate-transform-react/commit/9d1804c))
* **detection:** enhance error messages for asi errors in plain jsx ([048c407](https://github.com/sinnerschrader/patternplate-transform-react/commit/048c407))
* **detection:** resolve transitive demo dependencies properly ([e98bf39](https://github.com/sinnerschrader/patternplate-transform-react/commit/e98bf39))
* **detection:** throw better parser errors ([088d540](https://github.com/sinnerschrader/patternplate-transform-react/commit/088d540))
* **generation:** enable partial global support ([f6e72a7](https://github.com/sinnerschrader/patternplate-transform-react/commit/f6e72a7))
* **generation:** implement legacy global injection support ([1b92b54](https://github.com/sinnerschrader/patternplate-transform-react/commit/1b92b54))
* **transformation:** bootstrap ast-based react transformation ([f7f7b99](https://github.com/sinnerschrader/patternplate-transform-react/commit/f7f7b99))

marionebl committed:
Merge branch 'master' into feature/ast-detection-transformation
